File: basic-service-invocation/microservices/svc/wallet/handler/wallet.go
package handler

import (
	"context"
	crand "crypto/rand"
	"time"

	pb "github.com/kzmake/examples/basic-service-invocation/gen/go/wallet/v1"
	"github.com/oklog/ulid/v2"
)

type wallet struct {
	pb.UnimplementedWalletServiceServer
}

var _ pb.WalletServiceServer = new(wallet)

func NewWallet() pb.WalletServiceServer { return &wallet{} }

func (h *wallet) Create(ctx context.Context, req *pb.CreateRequest) (*pb.CreateResponse, error) {
	wallet := &pb.Wallet{
		Id: ulid.MustNew(ulid.Timestamp(time.Now()), crand.Reader).String(),
	}

	// do something

	return &pb.CreateResponse{Wallet: wallet}, nil
}

File: basic-service-invocation-py/microservices/svc/wallet/handler/wallet.py
import ulid
from grpclib.server import Stream

from basic.wallet.v1.wallet_grpc import WalletServiceBase
from basic.wallet.v1.wallet_pb2 import CreateRequest, CreateResponse, Wallet


class WalletService(WalletServiceBase):
    async def Create(self, stream: Stream[CreateRequest, CreateResponse]) -> None:
        req = await stream.recv_message()
        assert req is not None

        wallet = Wallet(id=str(ulid.new()))

        # do something

        res = CreateResponse(wallet=wallet)
        await stream.send_message(res)
